rust/oak/src/lib.rs

```diff
@@ -124,11 +123,40 @@ mod wasm {
     // See https://rustwasm.github.io/book/reference/js-ffi.html
     #[link(wasm_import_module = "oak")]
     extern "C" {
+        pub fn wait_on_channels(buf: *const u8, count: u32) -> i32;
         pub fn channel_read(handle: u64, buf: *mut u8, size: usize, actual_size: *mut u32) -> i32;
         pub fn channel_write(handle: u64, buf: *const u8, size: usize) -> i32;
     }
 }
 
+const SPACE_BYTES_PER_HANDLE: usize = 9;
+
+// Build a chunk of memory that is suitable for passing to wasm::wait_on_channels,
+// holding the given collection of channel handles.
+fn new_handle_space(handles: &[Handle]) -> Vec<u8> {
```

Comment:
Do you think it would be worth defining this in terms of a C struct with the appropriate fields, and link to that in the documentation? Similar to https://fuchsia.dev/fuchsia-src/zircon/syscalls/object_wait_many.md . Then I think you should be able to serialize / deserialize easily from both C++ and Rust by just reinterpreting bytes on both sides. Assuming C representation of structs is consistent in Wasm memory, but I think it is (please double check though).

Reply:
I was trying to avoid thinking about padding and alignment issues by specifying things on a definite byte-by-byte basis:
 - A C `struct` has no padding guarantees, so the layout for multiple consecutive instances could (theoretically) be compiler-dependent.
 - I don't immediately know how to get a block of bytes that has a specific alignment on the Rust/Wasm side.
 - I don't want to have C++ code that will fail in future if we start running on a new platform that is either big-endian, or which requires aligned access (ARM?).

(I've shifted to use [byteorder](https://docs.rs/byteorder/1.3.2/byteorder/) though, which should hopefully improve the chances of the generated Wasm code using `i64.store`)